docs/advanced-recipes/atom-creators.mdx

```diff
@@ -355,4 +355,105 @@ const ListOfThings = () => {
 }
 ```
 
-<CodeSandbox id="ugkzm0" />
+## atomWithDebounce
+
+> `atomWithDebounce` helps with creating an atom where state set should be debounced.
+
+This util is useful for text search inputs, where you would like to call **functions in derived atoms only once** after waiting for a duration, instead of firing an action on every keystroke.
+
+```tsx
+import { atom, Getter, PrimitiveAtom, Setter } from 'jotai'
+import { SetStateAction } from 'react'
+
+export default function atomWithDebounce<T>(
+  initialValue: T,
+  delayMilliseconds = 500,
+  shouldDebounceOnReset = false
+) {
+  const prevTimeoutAtom = atom<ReturnType<typeof setTimeout> | null>(null)
+
+  // DO NOT EXPORT currentValueAtom as using this atom to set state can cause
+  // inconsistent state between currentValueAtom and debouncedValueAtom
+  const _currentValueAtom = atom(initialValue)
+  const isDebouncingAtom = atom(false)
+
+  const clearTimeoutId = (
+    get: Getter,
+    set: Setter,
+    prevTimeoutAtom: PrimitiveAtom<ReturnType<typeof setTimeout> | null>
+  ) => {
+    const prevTimeoutId = get(prevTimeoutAtom)
+    if (prevTimeoutId) {
+      clearTimeout(prevTimeoutId)
+      set(prevTimeoutAtom, null)
+    }
+  }
+
+  const debouncedValueAtom = atom(
+    initialValue,
+    (get, set, update: SetStateAction<T>) => {
+      clearTimeoutId(get, set, prevTimeoutAtom)
+
+      const prevValue = get(_currentValueAtom)
+      const nextValue =
+        typeof update === 'function'
+          ? (update as (prev: T) => T)(prevValue)
+          : update
+
+      const onDebounceStart = () => {
+        set(_currentValueAtom, nextValue)
+        set(isDebouncingAtom, true)
+      }
+
+      const onDebounceEnd = () => {
+        set(debouncedValueAtom, nextValue)
+        set(isDebouncingAtom, false)
+      }
+
+      onDebounceStart()
+
+      if (!shouldDebounceOnReset && nextValue === initialValue) {
+        onDebounceEnd()
+        return
+      }
+
+      const nextTimeoutId = setTimeout(() => {
+        onDebounceEnd()
+      }, delayMilliseconds)
+
+      // set previous timeout atom in case it needs to get cleared
+      set(prevTimeoutAtom, nextTimeoutId)
+    }
+  )
+
+  // exported atom setter to clear timeout if needed
+  const clearTimeoutAtom = atom(null, (get, set, _arg) => {
+    clearTimeoutId(get, set, prevTimeoutAtom)
+    set(isDebouncingAtom, false)
+    set(_currentValueAtom, get(debouncedValueAtom))
```

Comment:
Why do you have this?
If we clear timeout, don't we want the old value, or do we want the new value?
Either way is fine for a recipe, though. Curious what would be expected in general.

Reply:
> If we clear timeout, don't we want the old value, or do we want the new value?

I misinterpreted AjaxSolution's comment as also to clear to the old value. Saw `useDebounce` implementation, and they simply just cancelled the debounce status and cleared the timeout. 

For now i will copy that implementation, and leave it to the users who read the docs to change the implementation to suit their usecase.

Reply:
The initial idea I had was to set it to the value before the debounce state was updated, so that "clearing" the debounce cycle would also mean resetting state to what it was before.

Reply:
[214c7aa](https://github.com/pmndrs/jotai/pull/1338/commits/214c7aa68d5ba263a8aa19b42802461383ad3f65)